gitlab/resource_gitlab_service_slack.go

```diff
@@ -0,0 +1,243 @@
+
+	opts := &gitlab.SetSlackServiceOptions{
+		WebHook: gitlab.String(d.Get("webhook").(string)),
+	}
+
+	opts.Username = gitlab.String(d.Get("username").(string))
+	opts.NotifyOnlyBrokenPipelines = gitlab.Bool(d.Get("notify_only_broken_pipelines").(bool))
+	opts.NotifyOnlyDefaultBranch = gitlab.Bool(d.Get("notify_only_default_branch").(bool))
+	opts.PushEvents = gitlab.Bool(d.Get("push_events").(bool))
+	opts.PushChannel = gitlab.String(d.Get("push_channel").(string))
+	opts.IssuesEvents = gitlab.Bool(d.Get("issues_events").(bool))
+	opts.IssueChannel = gitlab.String(d.Get("issue_channel").(string))
+	opts.ConfidentialIssuesEvents = gitlab.Bool(d.Get("confidential_issues_events").(bool))
+	opts.ConfidentialIssueChannel = gitlab.String(d.Get("confidential_issue_channel").(string))
+	opts.MergeRequestsEvents = gitlab.Bool(d.Get("merge_requests_events").(bool))
+	opts.MergeRequestChannel = gitlab.String(d.Get("merge_request_channel").(string))
+	opts.TagPushEvents = gitlab.Bool(d.Get("tag_push_events").(bool))
+	opts.TagPushChannel = gitlab.String(d.Get("tag_push_channel").(string))
+	opts.NoteEvents = gitlab.Bool(d.Get("note_events").(bool))
+	opts.NoteChannel = gitlab.String(d.Get("note_channel").(string))
+	opts.ConfidentialNoteEvents = gitlab.Bool(d.Get("confidential_note_events").(bool))
+	opts.PipelineEvents = gitlab.Bool(d.Get("pipeline_events").(bool))
+	opts.PipelineChannel = gitlab.String(d.Get("pipeline_channel").(string))
+	opts.WikiPageEvents = gitlab.Bool(d.Get("wiki_page_events").(bool))
+	opts.WikiPageChannel = gitlab.String(d.Get("wiki_page_channel").(string))
+
+	_, err := client.Services.SetSlackService(project, opts)
+	if err != nil {
+		return err
+	}
+
+	return resourceGitlabServiceSlackRead(d, meta)
+}
+
+func resourceGitlabServiceSlackRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+
+	log.Printf("[DEBUG] read gitlab slack service for project %s", project)
+
+	service, response, err := client.Services.GetSlackService(project)
+	if err != nil {
+		if response.StatusCode == 404 {
+			log.Printf("[WARN] removing Slack service from state because it no longer exists in Gitlab")
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	resourceGitlabServiceSlackSetToState(d, service)
+
+	return nil
+}
+
+func resourceGitlabServiceSlackUpdate(d *schema.ResourceData, meta interface{}) error {
+	return resourceGitlabServiceSlackCreate(d, meta)
+}
+
+func resourceGitlabServiceSlackDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+
+	log.Printf("[DEBUG] delete gitlab slack service for project %s", project)
+
+	_, err := client.Services.DeleteSlackService(project)
+	return err
+}
+
+func resourceGitlabServiceSlackImportState(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	d.Set("project", d.Id())
+
```

Comment:
shouldnt we fetch the service at this point from gitlab?

Reply:
Not sure what you mean here.

From what I understand, the `resourceGitlabServiceSlackImportState` function should just define the ID or necessary parameters so the READ function can then be called to actually fetch the resource's details.